Spec ModifyVolume for QoS kep-3751 (#544)
container-storage-interface/spec: spec.md

This RPC will be called by the CO to provision a new volume on behalf of a user (to be consumed as either a block device or a mounted filesystem).

This operation MUST be idempotent.
If a volume corresponding to the specified volume `name` already exists, is accessible from `accessibility_requirements`, and is compatible with the specified `capacity_range`, `volume_capabilities`, `parameters` and `mutable_parameters` in the `CreateVolumeRequest`, the Plugin MUST reply `0 OK` with the corresponding `CreateVolumeResponse`.

The `parameters` field SHALL contain opaque volume attributes to be specified at creation time. 

> The `mutable_parameters` field SHALL contain opaque volume attributes attributes that are defined at creation time but MAY also to be changed during the lifetime of the volume via a subsequent `ModifyVolume` RPC. 

Reviewer: ```suggestion
The `mutable_parameters` field SHALL contain opaque volume attributes that are defined at creation time but MAY also be changed during the lifetime of the volume via a subsequent `ControllerModifyVolume` RPC. 
```

> Values specified in `mutable_parameters` SHOULD take precedence over the values from `parameters`.

Reviewer: any reason in particular the wording is SHOULD vs MUST?

Author: There is no way to enforce MUST?

Reviewer: that doesn't stop us from using SHALL or MUST elsewhere...

Author: @bswartz How strong do you think our wording should be around this?

Reviewer 2: The only argument for SHOULD is that both fields are opaque and it's really hard to know what will go on inside the SP. That being said, I can't think of a specific example where an SP might NEED to violate this rule so I'm not opposed to using MUST instead. MUST definitely makes the expected semantic clearer on the CO side.

Author: Use MUST instead of SHOULD

This RPC will be called by the CO to provision a new volume on behalf of a user (to be consumed as either a block device or a mounted filesystem).

This operation MUST be idempotent.
If a volume corresponding to the specified volume `name` already exists, is accessible from `accessibility_requirements`, and is compatible with the specified `capacity_range`, `volume_capabilities`, `parameters` and `mutable_parameters` in the `CreateVolumeRequest`, the Plugin MUST reply `0 OK` with the corresponding `CreateVolumeResponse`.

The `parameters` field SHALL contain opaque volume attributes to be specified at creation time. 

> This field MUST be used to specify features of the volume that MUST NOT change. 

Reviewer 2: Just remove this whole line please

> This field SHOULD be used to specify features of the volume that SHOULD NOT change. 

Reviewer 2: This contradicts earlier guidance so we need to be careful here. SHOULD is gentler than MUST but possibly not gentle enough because many SPs are likely to continue allowing options which can change in this field forever as a backwards compatibility measure. Possibly just remove this line,

Author: Okay I changed it to:

```
This field MUST be used to specify features of the volume that MUST NOT change. 
```

This RPC will be called by the CO to provision a new volume on behalf of a user (to be consumed as either a block device or a mounted filesystem).

This operation MUST be idempotent.
If a volume corresponding to the specified volume `name` already exists, is accessible from `accessibility_requirements`, and is compatible with the specified `capacity_range`, `volume_capabilities`, `parameters` and `mutable_parameters` in the `CreateVolumeRequest`, the Plugin MUST reply `0 OK` with the corresponding `CreateVolumeResponse`.

> The `parameters` field allows certain volume attributes to be specified at creation time. This field SHOULD be used to specify features of the volume that can’t change. The `mutable_parameters` field allows other volume attributes to be defined both at creation time and also to be changed during the lifetime of the volume. Values specified in `mutable_parameters` SHOULD be applied after the values from `parameters`.

Reviewer 3: ```suggestion
The `parameters` field SHALL contain opaque volume attributes to be specified at creation time. This field SHOULD be used to specify features of the volume that SHOULD NOT change. The `mutable_parameters` field SHALL contain opaque volume attributes attributes that are defined at creation time but MAY also to be changed during the lifetime of the volume via a subsequent `ModifyVolume` RPC. Values specified in `mutable_parameters` SHOULD be applied after the values from `parameters`.
```

> The `parameters` field SHALL contain opaque volume attributes to be specified at creation time. This field SHOULD be used to specify features of the volume that SHOULD NOT change. The `mutable_parameters` field SHALL contain opaque volume attributes attributes that are defined at creation time but MAY also to be changed during the lifetime of the volume via a subsequent `ModifyVolume` RPC. Values specified in `mutable_parameters` SHOULD be applied after the values from `parameters`.

Reviewer: (1) one sentence per line, please

(2) "applied after" - see earlier comment re: alt wording

Author: (2) one sentence per line, do we need to add a new line between the sentences too?

Reviewer: > (2) one sentence per line, do we need to add a new line between the sentences too?

https://github.com/container-storage-interface/spec/blob/master/CONTRIBUTING.md#markdown-style

  // VOLUME_ACCESSIBILITY_CONSTRAINTS plugin capability, the SP MAY
  // choose where the provisioned volume is accessible from.
  TopologyRequirement accessibility_requirements = 7;

> 

Author: Changed it to:

```
  // Plugin specific parameters passed in as opaque key-value pairs for provision time.
  // This field is OPTIONAL. The Plugin is responsible for parsing and
  // validating these parameters. COs will treat these as opaque.
  map<string, string> parameters = 4;
```

  // This field is OPTIONAL. This allows the CO to specify the

>   // volume performance class parameters to apply. 

Reviewer 4: "volume performance class" is a K8s concept.   Change to "mutable volume parameters"?

Author: Modified

>   // volume performance class parameters to apply. 

Reviewer 2: I agree on not using K8s-specific terms, but we need specific and clear verbiage in the CreateVolume text that explains what this new field is for, and how it should be used, BOTH for the CO and the SP.

I am particularly worried about the potential for overlap between keys in the parameters and the mutable_parameters list. For example:
1) What should happen if a CO specifies a key for a mutable parameter in the existing parameters field?
2) What should happen if a CO specifies a key for an immutable parameter in the mutable parameters field?
3) What should happen if a CO specifies a key in both maps?
4) What should happen if there are different values associated with the two keys?
5) Is it valid for an SP to support the same key having different meanings in the different maps?

What I'm getting at is: how is having separate maps for existing parameters and mutable parameters functionally different from putting them in the same map? To be clear: I think they should be separate, because I think it gives us a better shot at detecting and reporting user errors, but we should think through what those errors are likely to look like, and how they will be caught and reported, and what kind of guidance will minimize the likelihood of problems.

>   map<string, string> mutable_parameters = 8;

Reviewer 4: Specify this as an alpha field.

  // VOLUME_ACCESSIBILITY_CONSTRAINTS plugin capability, the SP MAY
  // choose where the provisioned volume is accessible from.
  TopologyRequirement accessibility_requirements = 7;

>   // Plugins specific mutable parameters passed in as opaque key-value

Reviewer: ```suggestion
  // Plugin specific mutable parameters passed in as opaque key-value
```

  // pairs. This field is OPTIONAL. The Plugin is responsible for 
  // parsing and validating these parameters. COs will treat these 

>   // as opaque. Plugins SHOULD treat these as if they take precedence 

Reviewer: SHOULD vs MUST?

Reviewer 2: Same

>   // over the creation time parameters.

Reviewer: ```suggestion
  // over the parameters field.
```

>   map<string, string> mutable_parameters = 8 [(alpha_field) = true];

Reviewer: perhaps a note that indicates this mutable parameters field SHALL be ignored by the plugin if MUTABLE_VOLUME controller capability is not supported.

Reviewer 2: I agree with this but, I actually prefer the reverse of the language. We should say that the CO SHALL NOT provide anything in that field if the capability isn't advertised. We could even make both statements for double protection.

Author: I added:

```
// COs SHALL NOT provide any values in mutable_parameters if the
// capability is not enabled.
```

>   map<string, string> mutable_parameters = 8 [(alpha_field) = true];

Reviewer 5: Can there be a CSI driver which can't apply all `mutable_parameters` during volume creation but can only apply some of the other parameters via separate `ControllerModifyVolume` RPC call? 

In terms of integration with CO, this would mean specifying problematic `mutable_parameters` during volume creation could cause PV provisioning to fail. I am fine if we chose not to support this use case but I wonder if it is worth calling out,  so as something like csi-sanity tests could validate a driver for CSI spec compliance.

Reviewer 2: I don't think so. It's legal to call Modify right after Create, so an SP could just implement Create with mutable params as a create without mutable params followed immediately by a Modify (internally).

Requiring the CO so supply the mutable params at create time gives the SP a chance to avoid any expensive modifications by creating the volume correctly from the start. In theory we could have not included the param, but including it allows optimizations that wouldn't otherwise be possible. I see absolutely no downside from the SP perspective.

  // VOLUME_ACCESSIBILITY_CONSTRAINTS plugin capability, the SP MAY
  // choose where the provisioned volume is accessible from.
  TopologyRequirement accessibility_requirements = 7;

  // Plugin specific parameters to apply, passed in as opaque
  // key-value pairs. This field is OPTIONAL. The Plugin is

>   // responsible for  parsing and validating these parameters. 

Reviewer: ```suggestion
  // responsible for parsing and validating these parameters. 
```

Author: Good eyes.

  // This field is OPTIONAL. Refer to the `Secrets Requirements`
  // section on how to use this field.
  map<string, string> secrets = 5 [(csi_secret) = true];

  // See CreateVolumeRequest.mutable_parameters.
  // This field is OPTIONAL.

>   map<string, string> mutable_parameters = 6;

Reviewer: alpha field

Reviewer: ValidateVolumeCapabilitiesRequest

    // The volume creation parameters validated by the plugin.
    // This field is OPTIONAL.
    map<string, string> parameters = 3;

    // The volume creation mutable_parameters validated by the plugin.
    // This field is OPTIONAL.

>     map<string, string> mutable_parameters = 4;

Reviewer: alpha field

Reviewer 2: Good catch

Reviewer: ValidateVolumeCapabilitiesRequest

  google.protobuf.Int64Value iops = 1
  // The throughput of the volume is set. This field is OPTIONAL.
  google.protobuf.Int64Value throughput = 2
}
```

##### ModifyVolume Errors

| Condition | gRPC Code | Description | Recovery Behavior |
|-----------|-----------|-------------|-------------------|
| Exceeds capabilities | 3 INVALID_ARGUMENT | Indicates that the CO has specified capabilities not supported by the volume. | Caller MAY verify volume capabilities by calling ValidateVolumeCapabilities and retry with matching capabilities. |
| Volume does not exist | 5 NOT FOUND | Indicates that a volume corresponding to the specified volume_id does not exist. | Caller MUST verify that the volume_id is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |
| Volume in use | 9 FAILED_PRECONDITION | Indicates that the volume corresponding to the specified `volume_id` could not be updated because it is currently published on a node but the plugin does not have ONLINE expansion capability. | Caller SHOULD ensure that volume is not published and retry with exponential back off. |

> | Unsupported `capacity_range` | 11 OUT_OF_RANGE | Indicates that the capacity range is not allowed by the Plugin. More human-readable information MAY be provided in the gRPC `status.message` field. | Caller MUST fix the capacity range before retrying. |

Author: Currently the capacity_range is not listed in the ValidateVolumeCapabilities, I am not sure if I should add these to the ValidateVolumeCapabilities APIs but I will update this form.

> | Unsupported `capacity_range` | 11 OUT_OF_RANGE | Indicates that the capacity range is not allowed by the Plugin. More human-readable information MAY be provided in the gRPC `status.message` field. | Caller MUST fix the capacity range before retrying. |

Author: Oh i think you meant adding to message VolumeCapability, will add. Thanks!

|-----------|-----------|-------------|-------------------|
| Volume does not exist | 5 NOT_FOUND | Indicates that a volume corresponding to the specified `volume_id` does not exist. | Caller MUST verify that the `volume_id` is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |

#### `ModifyVolume`

> A Controller plugin SHALL implement this RPC call. If the MODIFY_VOLUME capability is present, a Controller plugin MUST implement this RPC call. This RPC allows the CO to change mutable key attributes of a volume. 

Reviewer 2: Canonical wording would be:

A Controller plugin MUST implement this RPC call if plugin has MODIFY_VOLUME controller capability.

> This operation MUST be idempotent. The preset parameters in the VolumePerformanceClass can be greater than or less than the current values. 

Reviewer 4: VolumePerformanceClass is a K8s term so it should not be in the CSI spec.
Suggest changing "The preset parameters in the VolumePerformanceClass can be greater than or less than the current values." to "The new mutable parameters in ModifyVolume can be different from the existing mutable parameters".

```protobuf
message ModifyVolumeRequest {
  option (alpha_message) = true;

  // Contains identity information for the existing volume.
  // This field is REQUIRED.
  string volume_id = 1;
  // This field is OPTIONAL.This allows the CO to specify the
  // mutable parameters to apply. 
  map<string, string> mutable_parameters = 2;
}

message ModifyVolumeResponse {
  option (alpha_message) = true;
}

```

##### ModifyVolume Errors

| Condition | gRPC Code | Description | Recovery Behavior |
|-----------|-----------|-------------|-------------------|

> | Parameters not supported | 3 INVALID_ARGUMENT | Indicates that the CO has specified volume performance parameters not supported by the volume. | Caller MAY verify volume performance parameters. |

Reviewer 4: "volume performance parameters" need to be updated.

Reviewer 2: Agree, and we should be clear that SPs SHOULD report errors if keys that are not mutable are specified. I don't think we can outright declare that any unknown parameter is an error, because that will cause problems for forward/backward compatibility, but certainly keys that are known not to be valid should trigger and error.

> | Exceeds capabilities | 3 INVALID_ARGUMENT | Indicates that the CO has specified capabilities not supported by the volume. | Caller MAY verify volume capabilities by calling ValidateVolumeCapabilities and retry with matching capabilities. |

Reviewer 4: Line 1659 and 1660 are two conditions with the same gRPC code?

Author: They are both INVALID_ARGUMENT, what other options do we have?

Reviewer 4: That's fine.  You can leave it as is.

|-----------|-----------|-------------|-------------------|
| Volume does not exist | 5 NOT_FOUND | Indicates that a volume corresponding to the specified `volume_id` does not exist. | Caller MUST verify that the `volume_id` is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |

> #### `ModifyVolume`

Reviewer: ```suggestion
#### `ControllerModifyVolume`
```

> A Controller plugin MUST implement this RPC call if plugin has MODIFY_VOLUME controller capability. This RPC allows the CO to change mutable key attributes of a volume. 

Reviewer: ```suggestion
A Controller plugin MUST implement this RPC call if plugin has MODIFY_VOLUME controller capability.
This RPC allows the CO to change mutable parameters of a volume. 
```

> This operation MUST be idempotent. The new mutable parameters in ModifyVolume can be different from the existing mutable parameters. 

Reviewer: ```suggestion
This operation MUST be idempotent.
The new mutable parameters in ControllerModifyVolume can be different from the existing mutable parameters. 
```

```protobuf
message ModifyVolumeRequest {
  option (alpha_message) = true;

  // Contains identity information for the existing volume.
  // This field is REQUIRED.
  string volume_id = 1;

  // Secrets required by plugin to complete modify volume request.
  // This field is OPTIONAL. Refer to the `Secrets Requirements`
  // section on how to use this field.
  map<string, string> secrets = 2 [(csi_secret) = true];

>   // Plugins specific parameters to apply, passed in as opaque key-value

Reviewer: ```suggestion
  // Plugin specific parameters to apply, passed in as opaque key-value
```

  // pairs. This field is OPTIONAL. The Plugin is responsible for
  // parsing and validating these parameters. COs will treat these
  // as opaque. COs MUST specify the intended value of every mutable
  // parameter. Absent values that were previously set MUST be

>   // interpretted as clearing the value.

Reviewer: "clearing"

.. meaning that the absent mutable parameter is considered *deleted* (vs set to empty string)?

Author: it will be considered as deleted.

Reviewer: Suggestion: Absent keys that were previously assigned a value MUST be interpreted as a deletion of the key from the set of mutable parameters.

  map<string, string> mutable_parameters = 3;
}

message ModifyVolumeResponse {
  option (alpha_message) = true;
}

```

##### ModifyVolume Errors

| Condition | gRPC Code | Description | Recovery Behavior |
|-----------|-----------|-------------|-------------------|
| Parameters not supported | 3 INVALID_ARGUMENT | Indicates that the CO has specified mutable parameters not supported by the volume. | Caller MAY verify mutable parameters. |

> | Exceeds capabilities | 3 INVALID_ARGUMENT | Indicates that the CO has specified capabilities not supported by the volume. | Caller MAY verify volume capabilities by calling ValidateVolumeCapabilities and retry with matching capabilities. |

Reviewer: ValidateVolumeCapabilitiesRequest has a `parameters` field but not a `mutable_parameters` field. why?

Author: I forget about it, we should add it here too.

|-----------|-----------|-------------|-------------------|
| Volume does not exist | 5 NOT_FOUND | Indicates that a volume corresponding to the specified `volume_id` does not exist. | Caller MUST verify that the `volume_id` is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |

#### `ModifyVolume`

A Controller plugin MUST implement this RPC call if plugin has MODIFY_VOLUME controller capability. This RPC allows the CO to change mutable key attributes of a volume. 

> SPs SHOULD NOT change attributes from `volume_context` in the `mutable_parameters`.

Reviewer 2: This is not the line I had in mind. I'm not sure what the best wording is, because someone would have to work really hard to misunderstand how volume_context works after reading its docs. But any new sentences we add need to go in the CreateVolume docs because that is where volume_context gets created.

Author: How about:

```
`volume_context` SHOULD NOT include any fields that is derived from `mutable_parameters`.  `mutable_parameters` and `volume_context` SHOULD be independent from each other.
```

Reviewer 2: I'm torn about whether we need any words at all, because there's no way to express the warning without restating information that's already clearly included in the spec. I'm going to review this one more time assuming no special wording is needed around mutable_parameters and volume_context.

Author: Removed the text from CreateVolume for now :)

A Controller plugin MUST implement this RPC call if plugin has MODIFY_VOLUME controller capability.
This RPC allows the CO to change mutable key attributes of a volume. 

This operation MUST be idempotent. 
The new mutable parameters in ControllerModifyVolume can be different from the existing mutable parameters. 

```protobuf
message ControllerModifyVolumeRequest {
  option (alpha_message) = true;

  // Contains identity information for the existing volume.
  // This field is REQUIRED.
  string volume_id = 1;

  // Secrets required by plugin to complete modify volume request.
  // This field is OPTIONAL. Refer to the `Secrets Requirements`
  // section on how to use this field.
  map<string, string> secrets = 2 [(csi_secret) = true];

  // Plugin specific parameters to apply, passed in as opaque key-value
  // pairs. This field is OPTIONAL. The Plugin is responsible for
  // parsing and validating these parameters. COs will treat these
  // as opaque. COs MUST specify the intended value of every mutable
  // parameter. Absent keys that were previously assigned a value MUST
  // be interpreted as a deletion of the key from the set of mutable
  // parameters.

>   map<string, string> mutable_parameters = 3;

Reviewer 5: > Absent keys that were previously assigned a value MUST
  // be interpreted as a deletion of the key from the set of mutable
  // parameters.

Does this mean that calling `ControllerModifyVolume` without specifying anything in `mutable_parameters` will mean, user is trying to reset these `mutable_parameters` to whatever default values driver/provider supports? What if storage provider does not support going back to default values?

It also feels like, this would generally mean driver should keep track of each `mutable_parameter` it supports and reset them if the keys are unspecified?  I am not sure, if this will be possible for each driver.

Reviewer 2: This wording was aimed mostly at the CO. I wanted to make it clear that if there are 2 mutable parameters "foo" and "bar" which were previously set, and the CO wants to modify "foo", it cannot just send a modify with the new value for "foo" and omit "bar", it needs to send the whole updated map of mutable parameters.

The goal here is to make it clear that the CO must implement a simple reconciler loop that always passes the whole map any time the map changes, and not something more complex where the CO computes a diff of what changed and sends only the diff.

Reviewer 5: From k8s perspective since values in `ControllerModifyVolume` will be coming from respective VAC class, this means that, if k8s admin wants a volume to be associated with new VAC class, they must specify all parameters from previous VAC class or CO should compute `parameters`  of `ControllerModifyVolume` by using parameters from both old and new VAC class.

Another problem I have is with wording `MUST be interpreted as deletion of key`. I don't know how driver will interpret this and what they are supposed to do.

Reviewer 2: If you have a better recommended wording I'm flexible here. I just recognized that in the previous spec wording that if you wanted to change two different values, you might have thought that you could do it as two separate modify calls each with one key in the map. I wanted to make it clear that the CO is supposed to send everything it knows each time, and not attempt to modify parameters one at a time.

I agree that it's not obvious what it means to delete/unset a value, and mandating that the SP does so not my intent. My point was that the CO should not omit values that it knows but does not intend to modify.

Reviewer 2: > From k8s perspective since values in `ControllerModifyVolume` will be coming from respective VAC class, this means that, if k8s admin wants a volume to be associated with new VAC class, they must specify all parameters from previous VAC class or CO should compute `parameters` of `ControllerModifyVolume` by using parameters from both old and new VAC class.

My intent is that if the VAC changes, the reconciler would just send the parameters for the new VAC. The contents of the old VAC would be irrelevant, so neither the CO nor the SP would need to know them in order for modify to complete successfully.

Reviewer 5: > My intent is that if the VAC changes, the reconciler would just send the parameters for the new VAC. The contents of the old VAC would be irrelevant, so neither the CO nor the SP would need to know them in order for modify to complete successfully.

It sounds like, what we want is to apply whatever parameter admin has provided in the VAC class and leave other parameters untouched. For example:

1. Volume gets created with no VAC class and lets say gets default iops of `10000` and throughput of `30`.
2. Admin associates the volume with VAC class `silver` , which has following properties:

```
iops: 20,0000
throughput: 100
```

So volume is modified accordingly.

3. Admin associates the volume with `gold` VAC class but somehow forgets to specify `throughput` parameter in VAC class. So `gold` VAC class has following properites:

```
iops: 75,0000
```

Now CO calls `ControllerModifyVolume` with `iops: 75,000` but this call does not reset the `throughput` to original value of `30` but leaves it at whatever last value was (which would be `100` from being attached to `silver` class).

If ^ flow is accurate then wording - `Absent keys that were previously assigned MUST be interpreted as a deletion` is problematic, because that would mean each VAC class should contain exhaustive list of parameters otherwise driver MUST somehow reset the original value of missing parameter.

If the intention is to apply parameters in `ControllerModifyVolume`  and ignore missing keys, then I am not sure why is `mutable_parameters` an `OPTIONAL` parameter in this call. It should not be optional and should not be empty.

But if the intention is indeed to reset the value of missing key to original value, then we have bunch of problems:

1.  This creates a error prone problem where each VAC admin creates MUST contain exhaustive list of key/value parameters, since any missing key will be interpreted as deletion and value has to be reset.
2. It creates a problem for driver authors, where they have to figure out - how to reset the value of missing key while  updating other parameters.

I would personally prefer to remove the wording about absent keys etc and make `mutable_parameters` non-optional and non-empty in `ControllerModifyVolume` RPC call. And add something like:

```
Plugin specific parameters to apply, passed in as opaque key-value
pairs. This field is REQUIRED and MUST not be empty. The Plugin is
responsible for parsing and validating these parameters. COs will
treat these as opaque. 
```

Reviewer 2: So the map always needs to optional for the reason that non-mutable parameters is optional -- because the map might be empty!

But I agree that the language around deletion might be too restrictive. I just didn't want to have a situation where CO authors thought they had the freedom to modify keys one at time when the user had requested multiple changes (i.e. change from one VAC to another). I want to communicate that the CO should send all of the changes it knows about in a single call, and I was looking for words that mean that.

Reviewer 5: So a `ControllerModifyVolume` call without `mutable_parameters` is a NO-OP right?

Reviewer 2: > So a `ControllerModifyVolume` call without `mutable_parameters` is a NO-OP right?

I don't think we can make that blanket statement. Some SPs will have mutable parameters that needs a mechanism to clear or reset them. If we define "key in map is empty during modify call" to mean no-op, then it become impossible to clear or reset one of these parameters.

I am okay with SP's choosing to interpret an omitted key as no change for SOME keys on SOME SPs, but we also need to allow a mechanism to reliably clear or reset mutable settings, and my proposal is to require the CO to always pass every parameter it knows, and to leave it up to the SP whether to interpret absence as a no-op or a clear. This makes the CO's job very easy, and gives the SP the power to implement a variety of things, with the burden of documentation falling on the specific vendor.

Reviewer 5: > to leave it up to the SP whether to interpret absence as a no-op or a clear.

That doesn't seem very consistent from API perspective. What is the use case of resetting the parameters? Can users not apply a different VAC to change or go back to previously used `mutable_parameters` ? At least in k8s APIs I think discussion is leading towards, not letting users set VAC back to nil once it has been set.

Reviewer 5: Also btw, the CSI spec has following wording about maps:

> A repeated or map field listed as REQUIRED MUST contain at least 1 element.

So we do seem to have some provision for making maps as REQUIRED. We haven't done that before though, so there is that. cc @jdef

Reviewer 2: Yeah if you define omitted keys as no-ops, then making the whole map optional makes no sense and the field should become required. I still think that there needs to be a way to unset previously-set mutable keys, which could be interpreted by the SP as reverting to some global default (defined elsewhere) or disabling some option (which was previously enabled).

Reviewer 5: > I still think that there needs to be a way to unset previously-set mutable keys, which could be interpreted by the SP as reverting to some global default (defined elsewhere) or disabling some option (which was previously enabled).

I was thinking of adding a `boolean` flag to accomplish this if needed but I think a cleaner approach might be to add a new RPC call. Something like - `ControllerResetVolumeModification` which does what you say. 

But for now IMO, I am not sure if we need it and we can always come back to it and add it in future. This allows us to keep `ControllerModifyVolume` RPC call relatively simpler.

Reviewer 2: So the things you suggest would also work, but they seem more complex. What's the downside of requiring the CO to pass all of the mutable options to each modify call and allowing the SP to interpret absent keys how it likes?

The point I'm trying to make here is that the CO implementation can be very simple and this detail doesn't matter (to the CO) because all the options are opaque anyways. The one thing I don't want to allow is updating a subset of the options in a single call, because I don't believe there is any benefit to allowing that.

Reviewer 6: With my Kubernetes hat on, I think it should not matter what VAC was set to a PVC before, important is only the  current VAC.

In other words, if a volume is created with VAC `bronze`, then modified to `silver` and then to `gold`, it should have the same QoS parameters as if it was created with `gold` right away. Regardless if any parameter was / was not dropped from the VACs. Otherwise it will be very difficult to see what QoS parameters a volume actually has. And current proposal here will allow this behavior with little / no code on CO side.

Yes, it puts more responsibility on cluster admins, they must count with resetting a missing parameter to default, or even always file all parameters to all VACs.

Reviewer 6: One idea: we could somehow clarify that `ControllerModifyVolume` should result in the same volume that would be created by `ControllerCreateVolume` with the same `parameters` as during original `ControllerCreateVolume` and new `mutable_parameters`. IMO that would be very clear to SP what to do, on the other hand they may need to store original `parameters` somewhere (in volume_context?), so they can return to them when corresponding key gets missing in `mutable_parameters`.

Author: That may not work for some SP's case, for example for GCP's hyperdisk-extreme, you can just specify the size of the disk, and a default size will be set for that disk even you did not specify one. But later you can actually modify the disk to have higher or lower iops as long as it is within the range.

}

message ControllerModifyVolumeResponse {
  option (alpha_message) = true;
}

```

> ##### ControllerModifyVolume Errors

Reviewer: i'm left wondering about mutable parameter changes that are not safe/possible to apply given the current state of the volume. e.g. the volume is node-published, and so tweaking parameter X is no longer possible in this state. however, it would be possible if the the volume were neither node-staged or node-published.

and if there are such cases, it would seem there's a need to talk about possible volume lifecycle dependencies somewhere (other than the error table for the RPC).

should there be a FAILED_PRECONDITION code for such cases? or else has this already been discussed?

Reviewer 2: This is a good catch, and it hadn't been discussed before. All of us were focused on the kinds of mutations that would be safe to do under any circumstances, but you're right that one could imagine some that were only safe to do on a unpublished volume.

Would you be opposed to limiting this feature to the kind of mutations that were always safe to do for volumes regardless of their published state? That would remove any need to discuss lifecycle here. If we permit the plugin to return FAILED_PRECONDITION due to the volume lifecycle it's going to be very hard on the CO side to compensate. The CO implementation of this feature is likely to be a fairly simple reconciler loop that's blind to the publish state of the volume. The most likely outcome of allowing the SP to return that error is that some volume just won't reconcile until the user detaches them and a constant stream of errors in the mean time.

Reviewer 7: It is probably imaginable that a change in parameters requires re-staging the volume to apply the parameters to a new connection (think of hostname change of an NFS-server). It may not be required immediately, but the change of the parameters only gets into effect once the volume has been disconnected and newly connected.

A `FAILED_PRECONDITION` is not sufficient in that case, I think. Maybe the `ControllerModifyVolumeResponse` could include an attribute to signal this?

Reviewer: interesting proposal. something like..
```protobuf
  // when unset, communicates to the CO that one or more mutable parameters
  // require a volume lifecycle state change before becoming effective.
  string awaiting_state_transition = N;
```
.. which could be even tighter w/ an enum for the blocking state transition

|-----------|-----------|-------------|-------------------|
| Volume does not exist | 5 NOT_FOUND | Indicates that a volume corresponding to the specified `volume_id` does not exist. | Caller MUST verify that the `volume_id` is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |

#### `ControllerModifyVolume`

> A Controller plugin MUST implement this RPC call if plugin has MODIFY_VOLUME controller capability.

Reviewer: ```suggestion
A Controller plugin MUST implement this RPC call if the plugin has the MODIFY_VOLUME controller capability.
```

This RPC allows the CO to change mutable key attributes of a volume. 

This operation MUST be idempotent. 

> The new mutable parameters in ControllerModifyVolume can be different from the existing mutable parameters. 

Reviewer 3: What does this mean?

Author: This means that, for example the existing mutable parameters in the volume has a field iops, and it is set to 500. The new mutable parameters in ControllerModifyVolume can set this field iops to something diff, like 1000.

```protobuf
message ControllerModifyVolumeRequest {
  option (alpha_message) = true;

  // Contains identity information for the existing volume.
  // This field is REQUIRED.
  string volume_id = 1;

  // Secrets required by plugin to complete modify volume request.
  // This field is OPTIONAL. Refer to the `Secrets Requirements`
  // section on how to use this field.
  map<string, string> secrets = 2 [(csi_secret) = true];

>   // Plugin specific parameters to apply, passed in as opaque key-value
>   // pairs. This field is REQUIRED. The Plugin is responsible for
>   // parsing and validating these parameters. COs will treat these
>   // as opaque. The CO SHOULD specify the intended values of all mutable
>   // parameters it intends to modify. SPs MUST NOT modify volumes based
>   // on the absence of keys, only keys that are specified should result
>   // in modifications to the volume.

Reviewer 3: ```suggestion  
  // Plugin specific volume attributes to mutate, passed in as opaque key-value pairs. 
  // This field is REQUIRED. The Plugin is responsible for
  // parsing and validating these parameters. COs will treat these
  // as opaque. The CO SHOULD specify the intended values of all mutable
  // parameters it intends to modify. SPs MUST NOT modify volumes based
  // on the absence of keys, only keys that are specified should result
  // in modifications to the volume.
```

  map<string, string> mutable_parameters = 3;
}

message ControllerModifyVolumeResponse {
  option (alpha_message) = true;
}

```

##### ControllerModifyVolume Errors

| Condition | gRPC Code | Description | Recovery Behavior |
|-----------|-----------|-------------|-------------------|
| Parameters not supported | 3 INVALID_ARGUMENT | Indicates that the CO has specified mutable parameters not supported by the volume. | Caller MAY verify mutable parameters. |
| Exceeds capabilities | 3 INVALID_ARGUMENT | Indicates that the CO has specified capabilities not supported by the volume. | Caller MAY verify volume capabilities by calling ValidateVolumeCapabilities and retry with matching capabilities. |

> | Volume does not exist | 5 NOT FOUND | Indicates that a volume corresponding to the specified volume_id does not exist. | Caller MUST verify that the volume_id is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |

Reviewer: ```suggestion
| Volume does not exist | 5 NOT_FOUND | Indicates that a volume corresponding to the specified volume_id does not exist. | Caller MUST verify that the volume_id is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |
```

.. and this looks like a bug several other places in the spec as well (and those others could be fixed in a separate PR)
